### PR TITLE
Re-write file location cache

### PR DESCRIPTION
**Pull request by qhz-ty**
1. Change crate name to `file_location_cache` to distinguish with the `gossip_cache` in `network` crate.
2. Garbage collect file announcements based on the origin `timestamp` in `SignedAnnounceFile` message.
3. Limit the total number of announcements instead of total files.

- - -
This change is [Reviewable](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58)

### CODE REVIEW

File: node/router/src/service.rs [hunk 1]
```diff
@@ -71,7 +67,7 @@ pub struct RouterService {
     store: Store,
 
     /// Cache for storing and serving gossip messages.
-    gossip_cache: Arc<GossipCache>,
+    gossip_cache: Arc<FileLocationCache>,
```
Comment (loganionion) [inline node/router/src/service.rs:70]: better to rename gossip_cache to file_location_cache?

File: node/router/src/service.rs [hunk 2]
```diff
@@ -88,7 +84,7 @@ impl RouterService {
         sync_send: SyncSender,
         miner_send: mpsc::UnboundedSender<MinerMessage>,
         store: Arc<dyn LogStore>,
-        gossip_cache: Arc<GossipCache>,
+        gossip_cache: Arc<FileLocationCache>,
```
Comment (loganionion) [inline node/router/src/service.rs:87]: same as above

File: node/src/client/builder.rs [hunk 3]
```diff
@@ -54,7 +54,7 @@ pub struct ClientBuilder {
     runtime_context: Option<RuntimeContext>,
     store: Option<Arc<dyn Store>>,
     async_store: Option<storage_async::Store>,
-    gossip_cache: Option<Arc<GossipCache>>,
+    gossip_cache: Option<Arc<FileLocationCache>>,
```
Comment (loganionion) [inline node/src/client/builder.rs:57]: rename

File: node/sync/src/controllers/serial.rs [hunk 4]
```diff
@@ -198,7 +193,7 @@ pub struct SerialSyncController {
     store: Store,
 
     /// Cache for storing and serving gossip messages.
-    gossip_cache: Arc<GossipCache>,
+    gossip_cache: Arc<FileLocationCache>,
```
Comment (loganionion) [inline node/sync/src/controllers/serial.rs:196]: rename

File: node/sync/src/controllers/serial.rs [hunk 5]
```diff
@@ -208,7 +203,7 @@ impl SerialSyncController {
         num_chunks: usize,
         ctx: Arc<SyncNetworkContext>,
         store: Store,
-        gossip_cache: Arc<GossipCache>,
+        gossip_cache: Arc<FileLocationCache>,
```
Comment (loganionion) [inline node/sync/src/controllers/serial.rs:206]: rename

File: node/sync/src/controllers/serial.rs [hunk 6]
```diff
@@ -229,7 +224,7 @@ impl SerialSyncController {
...
```
Comment (loganionion) [inline node/sync/src/controllers/serial.rs:227]: rename

File: node/sync/src/service.rs [hunk 7]
```diff
@@ -80,7 +80,7 @@ pub struct SyncService {
     store: Store,
 
     /// Cache for storing and serving gossip messages.
-    gossip_cache: Arc<GossipCache>,
+    gossip_cache: Arc<FileLocationCache>,
```
Comment (loganionion) [inline node/sync/src/service.rs:83]: rename

File: node/sync/src/service.rs [hunk 8]
```diff
@@ -95,7 +95,7 @@ impl SyncService {
         network_send: mpsc::UnboundedSender<NetworkMessage>,
         network_globals: Arc<NetworkGlobals>,
         store: Arc<dyn LogStore>,
-        gossip_cache: Arc<GossipCache>,
+        gossip_cache: Arc<FileLocationCache>,
```
Comment (loganionion) [inline node/sync/src/service.rs:98]: rename

**Review by loganionion — CHANGES_REQUESTED** [review]

**Review by qhz-ty — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58)* status: 0 of 18 files reviewed, 8 unresolved discussions (waiting on @ionian-dev, @loganionion, and @zz-ionian)

___
*[`node/router/src/service.rs` line 70 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2ZWNodOEKU__AN-r1-70:-N8vKXGHC_S9k6VrScJJ:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/router/src/service.rs#L70)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

better to rename gossip_cache to file_location_cache?
</blockquote></details>

Done.
___
*[`node/router/src/service.rs` line 87 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2ZWNodOEKU__AN-r1-87:-N8vKYtx7Ta6ypsaZRQJ:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/router/src/service.rs#L87)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`node/src/client/builder.rs` line 57 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2_XzRuhWfrLDKe-r1-57:-N8vK_UB9AgyY6F5lQDn:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/src/client/builder.rs#L57)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

rename
</blockquote></details>

Done.
___
*[`node/sync/src/service.rs` line 83 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2_XzRuhWfrLDKi-r1-83:-N8vKbcj6FJ7SF0Ix9YK:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/sync/src/service.rs#L83)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

rename
</blockquote></details>

Done.
___
*[`node/sync/src/service.rs` line 98 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2_XzRuhWfrLDKi-r1-98:-N8vKcCi9ssci5wwiskw:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/sync/src/service.rs#L98)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

rename
</blockquote></details>

Done.
___
*[`node/sync/src/controllers/serial.rs` line 196 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2_XzRuhWfrLDKg-r1-196:-N8vKd0X3jHHclfetmhL:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/sync/src/controllers/serial.rs#L196)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

rename
</blockquote></details>

Done.
___
*[`node/sync/src/controllers/serial.rs` line 206 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2_XzRuhWfrLDKg-r1-206:-N8vKesL6_uNAXF1Bc_3:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/sync/src/controllers/serial.rs#L206)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

rename
</blockquote></details>

Done.
___
*[`node/sync/src/controllers/serial.rs` line 227 at r1](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-N8h2g2_XzRuhWfrLDKg-r1-227:-N8vKfOP83FzAVHAfDit:b-896fix) ([raw file](https://github.com/ionian-web3-storage/ionian-rust/blob/0dd9017b865e5507d02a8669ca40e29a1a136c38/node/sync/src/controllers/serial.rs#L227)):*
<details><summary><i>Previously, loganionion wrote…</i></summary><blockquote>

rename
</blockquote></details>

Done.

**Review by loganionion — APPROVED** [review]
:lgtm:

> Reviewed 13 of 17 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ionian-web3-storage/ionian-rust/58#-:-N8vNBAD5EsLpQcp4ewn:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ionian-dev and @zz-ionian)

**Pull request merged**